File: Choonghun/개인 문제/3주차/프로그래머스 72413 - 합승 택시 요금/합승 택시 요금.py
def solution(n, s, a, b, fares):
    """
    n = 지점 개수
    s = 시작 지점
    a = A의 집
    b = B의 집
    """
    
    floyd = [[10e9 if i != j else 0 for j in range(n+1)] for i in range(n+1)]    
    for i,j,k in fares:
        floyd[i][j] = k
        floyd[j][i] = k
    
    for x in range(1,n+1):
        for y in range(1, n+1):
            for z in range(1, n+1):
                if floyd[y][z] > floyd[y][x] + floyd[x][z]:
                    floyd[y][z] = floyd[y][x] + floyd[x][z]
                    
    # print(floyd)
    answer = 10e9
    for i in range(1, n+1):
        answer = min(answer, floyd[s][i] + floyd[i][a] + floyd[i][b])
    
    return answer

File: Wonyoung/개인 문제/6주차/프로그래머스 72413 - 합승 택시 요금/합승 택시 요금.java
import java.util.*;

class Solution {
    public static int answer;
    public static int INF = Integer.MAX_VALUE;
    public static int[][] routes;
    // s: 출발 지점, a: 도착 지점(a), b: 도착 지점(b) 
    public int solution(int n, int s, int a, int b, int[][] fares) {
        answer = INF;
        routes = new int[n+1][n+1];
        
        for(int[] fare: fares){
            int c = fare[0], d = fare[1], f = fare[2];
            routes[c][d] = f;
            routes[d][c] = f;
        }
        
        int[] distS = dijkstra(s, n);
        int[] distA = dijkstra(a, n);
        int[] distB = dijkstra(b, n);
        
        for(int i = 1; i <= n; i++){
            int sx = distS[i];
            int xa = distA[i];
            int xb = distB[i];
            answer = Math.min(answer, sx + xa + xb);
        }

        return answer;
    }
    
    public static int[] dijkstra(int start, int n){
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt((int[] arr) -> arr[1]));
        int[] visited = new int[n+1];
        Arrays.fill(visited, INF);
        pq.add(new int[]{start, 0});
        visited[start] = 0;
            
        while(!pq.isEmpty()){
            int[] pos = pq.poll();
            int cur = pos[0], distance = pos[1];
            
            for(int next = 1; next <= n; next++){
                int cost = routes[cur][next];
                if(cost > 0 && visited[next] > distance + cost){
                    pq.add(new int[]{next, distance + cost});
                    visited[next] = distance + cost;
                }
            }
        }
                
        return visited;
    }
}